app/code/Magento/SalesRule/Model/Rule/DataProvider.php

```diff
@@ -60,17 +55,16 @@ public function __construct(
         $primaryFieldName,
         $requestFieldName,
         CollectionFactory $collectionFactory,
-        \Magento\Framework\Registry $registry,
-        \Magento\SalesRule\Model\Rule\Metadata\ValueProvider $metadataValueProvider,
+        Registry $registry,
+        protected readonly MetadataValueProvider $metadataValueProvider,
         array $meta = [],
         array $data = [],
-        DataPersistorInterface $dataPersistor = null
+        private ?DataPersistorInterface $dataPersistor = null
```

Comment:
The type of this argument seems to have changed by mistake.
```suggestion
        private DataPersistorInterface $dataPersistor = null,
```